README.md
# yahoofinancials
Welcome to Yahoo Financials!

Current Version: v0.10

Version Released: 08/14/2018

New Contributers Welcomed!

Email sandersconnor1@gmail.com with YAHOOFINANCIALS-{Your-Name} as email title if interested.

## Overview
A powerful financial data module used for pulling both fundamental and technical data from Yahoo Finance.
* As of Version 0.10, Yahoo Financials now returns historical pricing data for commodity futures, cryptocurrencies, ETFs, mutual funds, U.S. Treasuries, currencies, indexes, and stocks.

## Installation
* yahoofinancials runs fine on most versions of python 2 and 3.
* It was built and tested using versions 2.7 and 3.4
* The package depends on beautifulsoup4 and requests to work.

1. Installation using pip:
* Linux/Mac:
```R
$ pip install yahoofinancials
```
* Windows (If python doesn't work for you in cmd, try running the following command with just py):
```R
> python -m pip install yahoofinancials
```
2. Installation using github (Mac/Linux):
```R
$ git clone https://github.com/JECSand/yahoofinancials.git
$ cd yahoofinancials
$ python setup.py install
```
3. Demo using the included demo script:
```R
$ cd yahoofinancials
$ python demo.py -h
$ python demo.py
$ python demo.py WFC C BAC
```

## Module Methods
* The financial data from all methods is returned as JSON.
* You can run multiple symbols at once using an inputted array or run an individual symbol using an inputted string.
* YahooFinancials works on most versions of python 2 and 3 and runs on all operating systems. (Windows, Mac, Linux).

### Featured Methods
1. get_financial_stmts(frequency, statement_type, reformat=True)
   * frequency can be either 'annual' or 'quarterly'.
   * statement_type can be 'income', 'balance', 'cash' or a list of several.
   * reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
2. get_stock_price_data(reformat=True)
   * reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
3. get_stock_earnings_data(reformat=True)
   * reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
4. get_summary_data(reformat=True)
   * New in v0.10.
   * Returns financial summary data for cryptocurrencies, stocks, currencies, ETFs, mutual funds, U.S. Treasuries, commodity futures, and indexes.
   * reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
5. get_stock_quote_type_data()
6. get_historical_price_data(start_date, end_date, time_interval)
   * New in v0.9.
   * This method will pull historical pricing data for stocks, currencies, ETFs, mutual funds, U.S. Treasuries, cryptocurrencies, commodities, and indexes.
   * start_date should be entered in the 'YYYY-MM-DD' format and is the first day that data will be pulled for.
   * end_date should be entered in the 'YYYY-MM-DD' format and is the last day that data will be pulled for.
   * time_interval can be either 'daily', 'weekly', or 'monthly'. This variable determines the time period interval for your pull.
   * Data response includes relevant pricing event data such as dividends and stock splits.
7. get_num_shares_outstanding(price_type='current'):
   * price_type can also be set to 'average' to calculate the shares outstanding with the daily average price.

### Methods Depreciated in v0.9 & v0.10
* get_stock_summary_data():
  - Scheduled for removal in v1.0
* get_historical_stock_data():
  - Scheduled for removal in v1.0

### Additional Module Methods
* get_interest_expense():
* get_operating_income():
* get_total_operating_expense():
* get_total_revenue():
* get_cost_of_revenue():
* get_income_before_tax():
* get_income_tax_expense():
* get_gross_profit():
* get_net_income_from_continuing_ops():
* get_research_and_development():
* get_current_price():
* get_current_change():
* get_current_percent_change():
* get_current_volume():
* get_prev_close_price():
* get_open_price():
* get_ten_day_avg_daily_volume():
* get_three_month_avg_daily_volume():
* get_stock_exchange():
* get_market_cap():
* get_daily_low():
* get_daily_high():
* get_currency():
* get_yearly_high():
* get_yearly_low():
* get_dividend_yield():
* get_annual_avg_div_yield():
* get_five_yr_avg_div_yield():
* get_dividend_rate():
* get_annual_avg_div_rate():
* get_50day_moving_avg():
* get_200day_moving_avg():
* get_beta():
* get_payout_ratio():
* get_pe_ratio():
* get_price_to_sales():
* get_exdividend_date():
* get_book_value():
* get_ebit():
* get_net_income():
* get_earnings_per_share():

## Usage Examples
* The class constructor can take either a single ticker or a list of tickers as it's parameter.
* This makes it easy to initiate multiple classes for different groupings of financial assets.
* Quarterly statement data returns the last 4 periods of data, while annual returns the last 3.

### Single Ticker Example
```R
from yahoofinancials import YahooFinancials

ticker = 'AAPL'
yahoo_financials = YahooFinancials(ticker)

balance_sheet_data_qt = yahoo_financials.get_financial_stmts('quarterly', 'balance')
income_statement_data_qt = yahoo_financials.get_financial_stmts('quarterly', 'income')
all_statement_data_qt =  yahoo_financials.get_financial_stmts('quarterly', ['income', 'cash', 'balance'])
apple_earnings_data = yahoo_financials.get_stock_earnings_data()
apple_net_income = yahoo_financials.get_net_income()
historical_stock_prices = yahoo_financials.get_historical_price_data('2015-01-15', '2017-10-15', 'weekly')
```

### Lists of Tickers Example
```R
from yahoofinancials import YahooFinancials

tech_stocks = ['AAPL', 'MSFT', 'INTC']
bank_stocks = ['WFC', 'BAC', 'C']
commodity_futures = ['GC=F', 'SI=F', 'CL=F']
cryptocurrencies = ['BTC-USD', 'ETH-USD', 'XRP-USD']
currencies = ['EURUSD=X', 'JPY=X', 'GBPUSD=X']
mutual_funds = ['PRLAX', 'QASGX', 'HISFX']
us_treasuries = ['^TNX', '^IRX', '^TYX']

yahoo_financials_tech = YahooFinancials(tech_stocks)
yahoo_financials_banks = YahooFinancials(bank_stocks)
yahoo_financials_commodities = YahooFinancials(commodity_futures)
yahoo_financials_cryptocurrencies = YahooFinancials(cryptocurrencies)
yahoo_financials_currencies = YahooFinancials(currencies)
yahoo_financials_mutualfunds = YahooFinancials(mutual_funds)
yahoo_financials_treasuries = YahooFinancials(us_treasuries)

tech_cash_flow_data_an = yahoo_financials_tech.get_financial_stmts('annual', 'cash')
bank_cash_flow_data_an = yahoo_financials_banks.get_financial_stmts('annual', 'cash')

banks_net_ebit = yahoo_financials_banks.get_ebit()
tech_stock_price_data = yahoo_financials_tech.get_stock_price_data()
daily_bank_stock_prices = yahoo_financials_banks.get_historical_price_data('2008-09-15', '2017-09-15', 'daily')
daily_commodity_prices = yahoo_financials_commodities.get_historical_price_data('2008-09-15', '2017-09-15', 'daily')
daily_crypto_prices = yahoo_financials_cryptocurrencies.get_historical_price_data('2008-09-15', '2017-09-15', 'daily')
daily_currency_prices = yahoo_financials_currencies.get_historical_price_data('2008-09-15', '2017-09-15', 'daily')
daily_mutualfund_prices = yahoo_financials_mutualfunds.get_historical_price_data('2008-09-15', '2017-09-15', 'daily')
daily_treasury_prices = yahoo_financials_treasuries.get_historical_price_data('2008-09-15', '2017-09-15', 'daily')
```

## Examples of Returned JSON Data
1. Annual Income Statement Data for Apple:
```R
yahoo_financials = YahooFinancials('AAPL')
print(yahoo_financials.get_financial_stmts('annual', 'income'))
```
```javascript
{
    "incomeStatementHistory": {
        "AAPL": [
            {
                "2016-09-24": {
                    "minorityInterest": null,
                    "otherOperatingExpenses": null,
                    "netIncomeFromContinuingOps": 45687000000,
                    "totalRevenue": 215639000000,
                    "totalOtherIncomeExpenseNet": 1348000000,
                    "discontinuedOperations": null,
                    "incomeTaxExpense": 15685000000,
                    "extraordinaryItems": null,
                    "grossProfit": 84263000000,
                    "netIncome": 45687000000,
                    "sellingGeneralAdministrative": 14194000000,
                    "interestExpense": null,
                    "costOfRevenue": 131376000000,
                    "researchDevelopment": 10045000000,
                    "netIncomeApplicableToCommonShares": 45687000000,
                    "effectOfAccountingCharges": null,
                    "incomeBeforeTax": 61372000000,
                    "otherItems": null,
                    "operatingIncome": 60024000000,
                    "ebit": 61372000000,
                    "nonRecurring": null,
                    "totalOperatingExpenses": 0
                }
            }
        ]
    }
}
```
2. Annual Balance Sheet Data for Apple:
```R
yahoo_financials = YahooFinancials('AAPL')
print(yahoo_financials.get_financial_stmts('annual', 'balance'))
```
```javascript
{
    "balanceSheetHistory": {
        "AAPL": [
            {
                "2016-09-24": {
                    "otherCurrentLiab": 8080000000,
                    "otherCurrentAssets": 8283000000,
                    "goodWill": 5414000000,
                    "shortTermInvestments": 46671000000,
                    "longTermInvestments": 170430000000,
                    "cash": 20484000000,
                    "netTangibleAssets": 119629000000,
                    "totalAssets": 321686000000,
                    "otherLiab": 36074000000,
                    "totalStockholderEquity": 128249000000,
                    "inventory": 2132000000,
                    "retainedEarnings": 96364000000,
                    "intangibleAssets": 3206000000,
                    "totalCurrentAssets": 106869000000,
                    "otherStockholderEquity": 634000000,
                    "shortLongTermDebt": 11605000000,
                    "propertyPlantEquipment": 27010000000,
                    "deferredLongTermLiab": 2930000000,
                    "netReceivables": 29299000000,
                    "otherAssets": 8757000000,
                    "longTermDebt": 75427000000,
                    "totalLiab": 193437000000,
                    "commonStock": 31251000000,
                    "accountsPayable": 59321000000,
                    "totalCurrentLiabilities": 79006000000
                }
            }
        ]
    }
}
```
3. Quarterly Cash Flow Statement Data for Citigroup:
```R
yahoo_financials = YahooFinancials('C')
print(yahoo_financials.get_financial_stmts('quarterly', 'cash'))
```
```javascript
{
    "cashflowStatementHistoryQuarterly": {
        "C": [
            {
                "2017-06-30": {
                    "totalCashFromOperatingActivities": -18505000000,
                    "effectOfExchangeRate": -117000000,
                    "totalCashFromFinancingActivities": 39798000000,
                    "netIncome": 3872000000,
                    "dividendsPaid": -760000000,
                    "salePurchaseOfStock": -1781000000,
                    "capitalExpenditures": -861000000,
                    "changeToLiabilities": -7626000000,
                    "otherCashflowsFromInvestingActivities": 82000000,
                    "totalCashflowsFromInvestingActivities": -22508000000,
                    "netBorrowings": 33586000000,
                    "depreciation": 901000000,
                    "changeInCash": -1332000000,
                    "changeToNetincome": 1444000000,
                    "otherCashflowsFromFinancingActivities": 8753000000,
                    "changeToOperatingActivities": -17096000000,
                    "investments": -23224000000
                }
            }
        ]
    }
}
```
4. Monthly Historical Stock Price Data for Wells Fargo:
```R
yahoo_financials = YahooFinancials('WFC')
print(yahoo_financials.get_historical_price_data("2018-07-10", "2018-08-10", "monthly"))
```
```javascript
{
    "WFC": {
        "currency": "USD",
        "eventsData": {
            "dividends": {
                "2018-08-01": {
                    "amount": 0.43,
                    "date": 1533821400,
                    "formatted_date": "2018-08-09"
                }
            }
        },
        "firstTradeDate": {
            "date": 76233600,
            "formatted_date": "1972-06-01"
        },
        "instrumentType": "EQUITY",
        "prices": [
            {
                "adjclose": 57.19147872924805,
                "close": 57.61000061035156,
                "date": 1533096000,
                "formatted_date": "2018-08-01",
                "high": 59.5,
                "low": 57.08000183105469,
                "open": 57.959999084472656,
                "volume": 138922900
            }
        ],
        "timeZone": {
            "gmtOffset": -14400
        }
    }
}
```
5. Monthly Historical Price Data for EURUSD:
```R
yahoo_financials = YahooFinancials('EURUSD=X')
print(yahoo_financials.get_historical_price_data("2018-07-10", "2018-08-10", "monthly"))
```
```javascript
{
    "EURUSD=X": {
        "currency": "USD",
        "eventsData": {},
        "firstTradeDate": {
            "date": 1070236800,
            "formatted_date": "2003-12-01"
        },
        "instrumentType": "CURRENCY",
        "prices": [
            {
                "adjclose": 1.1394712924957275,
                "close": 1.1394712924957275,
                "date": 1533078000,
                "formatted_date": "2018-07-31",
                "high": 1.169864296913147,
                "low": 1.1365960836410522,
                "open": 1.168961763381958,
                "volume": 0
            }
        ],
        "timeZone": {
            "gmtOffset": 3600
        }
    }
}
```
6. Monthly Historical Price Data for BTC-USD:
```R
yahoo_financials = YahooFinancials('BTC-USD')
print(yahoo_financials.get_historical_price_data("2018-07-10", "2018-08-10", "monthly"))
```
```javascript
{
    "BTC-USD": {
        "currency": "USD",
        "eventsData": {},
        "firstTradeDate": {
            "date": 1279321200,
            "formatted_date": "2010-07-16"
        },
        "instrumentType": "CRYPTOCURRENCY",
        "prices": [
            {
                "adjclose": 6285.02001953125,
                "close": 6285.02001953125,
                "date": 1533078000,
                "formatted_date": "2018-07-31",
                "high": 7760.740234375,
                "low": 6133.02978515625,
                "open": 7736.25,
                "volume": 4334347882
            }
        ],
        "timeZone": {
            "gmtOffset": 3600
        }
    }
}
```
7. Weekly Historical Price Data for Crude Oil Futures:
```R
yahoo_financials = YahooFinancials('CL=F')

CHANGES
0.1   10/13/2017 -- Initial release.
0.2   10/20/2017 -- Added New Methods and Restructured Module Classes
0.3   10/25/2017 -- Added New Methods and Calculated Measures
0.4   03/06/2018 -- Fixed reported bug on line 470 of init.py by adding a try and except with suggested line of code.
0.6   07/06/2018 -- Fixed reported bug in format_date function.
0.6   07/06/2018 -- Added exception handler that returns a ticker with no available data with a null value.
0.7   08/03/2018 -- Merged Slyvandb's improvements into the master branch.
0.7   08/03/2018 -- Added a try catch at line 465 to explicitly type cast the dict keys to a list if the initial attempt fails.
0.7   08/03/2018 -- Added 10 new income statement history methods beginning at line 567.
0.7   08/03/2018 -- Added a fix for trevorwelch's open issue involving the unnecessary sys.exit(1) on line 286 by replacing it with return None.
0.8   08/11/2018 -- Added a new method to get the current day's shares outstanding called get_num_shares_outstanding() starting on line 617.
0.9   08/14/2018 -- Added a new method called get_historical_price_data() to get price data for commodity futures, indexes, currencies, and cryptos in addition to stocks.
0.9   08/14/2018 -- Depreciated the get_historical_stock_data() method and scheduled it's removal for version 1.0.
0.10  08/14/2018 -- Added a new Method to get summary data for stocks, indexes, cryptocurrencies, currencies, and commodity futures, get_summary_data().
0.10  08/14/2018 -- Depreciated the get_stock_summary_data() method and scheduled it's removal for version 1.0.
1.0   08/22/2018 -- Removed the get_historical_stock_data() method.
1.0   08/22/2018 -- Removed the get_stock_summary_data() method.
1.0   08/22/2018 -- Removed the requests dependency and replaced it with urllib.
1.0   08/22/2018 -- Updated README.md to README.rst
1.1   08/23/2018 -- Fixed net income python 3 error
1.2   08/29/2018 -- Fixed timezone error in python testing. Issue #11.
1.2   08/29/2018 -- Fixed unicode formatted_date string for consistency.
1.3   10/25/2018 -- Added patch for reported bug described in Issue #14.
1.4   01/13/2019 -- Python3.7 support added.
1.4   01/13/2019 -- Added patch for reported bug described in Issue #15.
1.4   01/13/2019 -- Added method for get_stock_earnings_data().
1.5   01/27/2019 -- Added get_daily_dividend_data() method as request in Issue #20.
1.5   01/27/2019 -- Added test_yf_dividend_price() unit testing method.
1.6   10/18/2020 -- Merged in two pull requests with bug fixes from sylvandb.
1.7   01/01/2023 -- Merged in pull request with fixes from sedwards2000.
1.7   01/01/2023 -- Support for Python 2.7 dropped.
1.8   01/09/2023 -- Improved decryption support.
1.9   01/14/2023 -- Added official Python 3.11 support.
1.9   01/14/2023 -- General code cleanup & quality improvements.
1.9   01/14/2023 -- Added new optional concurrent input to YahooFinancials(), if True extraction will run async.
1.9   01/14/2023 -- Added new optional country input to YahooFinancials(), currently supports language & region.
1.9   01/14/2023 -- Added new optional proxies input to YahooFinancials(), currently supports a list or string.
1.9   01/14/2023 -- Updated beautifulsoup4 find to use string parameter instead deprecated text parameter.
1.9   01/14/2023 -- Replace urllib with Requests due to deprecation warning.
1.9   01/14/2023 -- Merged in pull request from shaunpatterson.
1.9   01/14/2023 -- Merged in pull request from flipdazed.
1.9   01/14/2023 -- Fixed new data encryption issue & hardened.
1.10  01/25/2023 -- Fixed new decryption issue.
1.11  01/26/2023 -- Added a dynamic fix for the decryption issue.
1.12  01/27/2023 -- Fixed get profile function for #127 and added additional unit test.
1.13  02/14/2023 -- Implemented fixes for #132 and #128 by refactoring package to use Yahoo API instead of scraping.
1.13  02/14/2023 -- Added method to retrieve ESG data as requested in #48.
1.13  02/14/2023 -- Added additional unit tests.
1.14  02/21/2023 -- Fixed get_ten_day_avg_daily_volume as reported in #137.
1.14  02/21/2023 -- Removed get_three_month_avg_daily_volume due to value now missing in Yahoo data.
1.14  02/21/2023 -- Added unit test for get_ten_day_avg_daily_volume.
1.15  06/21/2023 -- Dropped Python 3.6 support.
1.15  06/21/2023 -- Enhanced cache to cover all api requests.
1.15  06/21/2023 -- Enhanced api url to better toggle between query1 and query2 subdomains.
1.15  06/21/2023 -- Minimized sleeps in between 400 errors to less than 10 seconds.
1.16  07/17/2023 -- Minor code cleanup in test_yahoofinancials.py.
1.16  07/17/2023 -- Fixed HTTP 401 error when executing get_current_price() as reported in #152.
1.17  10/30/2023 -- Added a fix for 404 error reported in #157.
1.18  12/09/2023 -- Merged in branch from shaunpatterson to fix #164.
1.19  12/12/2023 -- Refactored session management system to handle cookie and crumbs better.
1.19  12/12/2023 -- Added fixes for #167, #166, #160.
1.20  12/16/2023 -- Merged in pull request #171 from bjosun.
1.20  12/17/2023 -- Added optional flat format output param on YahooFinancial class.
1.20  12/17/2023 -- Added get_insight() and get_recommendations() methods.


yahoofinancials/yf.py
"""
==============================
The Yahoo Financials Module
Version: 1.20
==============================

Author: Connor Sanders
Email: jecsand@pm.me
Version Released: 12/17/2023
Tested on Python 3.7, 3.8, 3.9, 3.10, 3.11, and 3.12

Copyright (c) 2023 Connor Sanders
MIT License

List of Included Functions:

1) get_financial_stmts(frequency, statement_type, reformat=True)
   - frequency can be either 'annual' or 'quarterly'.
   - statement_type can be 'income', 'balance', 'cash'.
   - reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
2) get_stock_price_data(reformat=True)
3) get_stock_earnings_data()
   - reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
4) get_summary_data(reformat=True)
   - reformat optional value defaulted to true. Enter False for unprocessed raw data from Yahoo Finance.
5) get_stock_quote_type_data()
6) get_historical_price_data(start_date, end_date, time_interval)
   - Gets historical price data for currencies, stocks, indexes, cryptocurrencies, and commodity futures.
   - start_date should be entered in the 'YYYY-MM-DD' format. First day that financial data will be pulled.
   - end_date should be entered in the 'YYYY-MM-DD' format. Last day that financial data will be pulled.
   - time_interval can be either 'daily', 'weekly', or 'monthly'. Parameter determines the time period interval.

Usage Examples:
from yahoofinancials import YahooFinancials
#tickers = 'AAPL'
#or
tickers = ['AAPL', 'WFC', 'F', 'JPY=X', 'XRP-USD', 'GC=F']
yahoo_financials = YahooFinancials(tickers)
balance_sheet_data = yahoo_financials.get_financial_stmts('quarterly', 'balance')
earnings_data = yahoo_financials.get_stock_earnings_data()
historical_prices = yahoo_financials.get_historical_price_data('2015-01-15', '2017-10-15', 'weekly')
"""

from yahoofinancials.calcs import num_shares_outstanding, eps
from yahoofinancials.data import YahooFinanceData

__version__ = "1.20"
__author__ = "Connor Sanders"


# Class containing methods to create stock data extracts
class YahooFinancials(YahooFinanceData):
    """
    Arguments
    ----------
    tickers: str or list
        Ticker or listed collection of tickers
    Keyword Arguments
    -----------------
    concurrent: bool, default False, optional
        Defines whether the requests are made synchronously or asynchronously.
    country: str, default 'US', optional
        This allows you to alter the region, lang, corsDomain parameter sent with each request based on selected country
    max_workers: int, default 8, optional
        Defines the number of workers used to make concurrent requests.
        Only relevant if concurrent=True
    timeout: int, default 30, optional
        Defines how long a request will stay open.
    proxies: str or list, default None, optional
        Defines any proxies to use during this instantiation.
    flat_format: bool, default False, optional
        If set to True, returns fundamental data in a flattened format, i.e. without the list of dicts.
    """

    # Private method that handles financial statement extraction
    def _run_financial_stmt(self, statement_type, report_num, frequency, reformat):
        hist_obj = {"interval": frequency}
        report_name = self.YAHOO_FINANCIAL_TYPES[statement_type][report_num]
        if reformat:
            raw_data = self.get_stock_data(statement_type, report_name=report_name, hist_obj=hist_obj)
            data = self.get_reformatted_stmt_data(raw_data)
        else:
            data = self.get_stock_data(statement_type, report_name=report_name, hist_obj=hist_obj)
        return data

    # Public Method for the user to get financial statement data
    def get_financial_stmts(self, frequency, statement_type, reformat=True):
        report_num = self.get_report_type(frequency)
        if isinstance(statement_type, str):
            data = self._run_financial_stmt(statement_type, report_num, frequency, reformat)
        else:
            data = {}
            for stmt_type in statement_type:
                re_data = self._run_financial_stmt(stmt_type, report_num, frequency, reformat)
                data.update(re_data)
        return data

    # Public Method for the user to get stock price data
    def get_stock_price_data(self, reformat=True):
        if reformat:
            return self.get_clean_data(self.get_stock_tech_data('price'), 'price')
        else:
            return self.get_stock_tech_data('price')

    # Public Method for the user to return key-statistics data
    def get_key_statistics_data(self, reformat=True):
        if reformat:
            return self.get_clean_data(self.get_stock_tech_data('defaultKeyStatistics'), 'defaultKeyStatistics')
        else:
            return self.get_stock_tech_data('defaultKeyStatistics')

    # Public Method for the user to get company profile data
    def get_stock_profile_data(self, reformat=True):
        if reformat:
            return self.get_clean_data(
                self.get_stock_data(statement_type='profile', tech_type='assetProfile', report_name='assetProfile'),
                'earnings')
        else:
            return self.get_stock_data(statement_type='profile', tech_type='assetProfile', report_name='assetProfile')

    # Public Method for the user to get stock earnings data
    def get_stock_earnings_data(self):
        return self.get_stock_tech_data('earnings')

    # Public Method for the user to return financial data
    def get_financial_data(self, reformat=True):
        if reformat:
            return self.get_clean_data(self.get_stock_data(statement_type='keystats', tech_type='financialData'),
                                       'financialData')
        else:
            return self.get_stock_data(statement_type='keystats', tech_type='financialData')

    # Public Method for the user to get stock summary data
    def get_summary_data(self, reformat=True):
        if reformat:
            return self.get_clean_data(self.get_stock_tech_data('summaryDetail'), 'summaryDetail')
        else:
            return self.get_stock_tech_data('summaryDetail')

    # Public Method for the user to get the yahoo summary url
    def get_stock_summary_url(self):
        if isinstance(self.ticker, str):
            return self._BASE_YAHOO_URL + self.ticker
        return {t: self._BASE_YAHOO_URL + t for t in self.ticker}

    # Public Method for the user to get stock quote data
    def get_stock_quote_type_data(self):
        return self.get_stock_tech_data('quoteType')

    # Public Method for the user to get stock quote data
    def get_esg_score_data(self):
        return self.get_stock_tech_data('esgScores')

    def _get_analytic_data(self, tech_type):
        return self.get_stock_data(statement_type='analytic', tech_type=tech_type)

    # Public Method for user to get historical price data with
    def get_historical_price_data(self, start_date, end_date, time_interval):
        interval_code = self.get_time_code(time_interval)
        start = self.format_date(start_date)
        end = self.format_date(end_date)
        hist_obj = {'start': start, 'end': end, 'interval': interval_code}
        return self.get_stock_data('history', hist_obj=hist_obj)

    # Private Method for Functions needing stock_price_data
    def _stock_price_data(self, data_field):
        price_data = self.get_stock_price_data()
        if isinstance(self.ticker, str):
            if price_data[self.ticker] is None:
                return None
            return price_data[self.ticker].get(data_field)
        else:
            ret_obj = {}
            for tick in self.ticker:
                if price_data[tick] is None:
                    ret_obj.update({tick: None})
                else:
                    ret_obj.update({tick: price_data[tick].get(data_field)})
            return ret_obj

    # Private Method for Functions needing stock_price_data
    def _stock_summary_data(self, data_field):
        sum_data = self.get_summary_data()
        if isinstance(self.ticker, str):
            if sum_data[self.ticker] is None:
                return None
            return sum_data[self.ticker].get(data_field, None)
        else:
            ret_obj = {}
            for tick in self.ticker:
                if sum_data[tick] is None:
                    ret_obj.update({tick: None})
                else:
                    ret_obj.update({tick: sum_data[tick].get(data_field, None)})
            return ret_obj

    # Private Method for Functions needing financial statement data
    def _financial_statement_data(self, stmt_type, stmt_code, field_name, freq):
        re_data = self.get_financial_stmts(freq, stmt_type)[stmt_code]
        if isinstance(self.ticker, str):
            try:
                date_key = re_data[self.ticker][0].keys()[0]
            except (IndexError, AttributeError, TypeError):
                date_key = list(re_data[self.ticker][0])[0]
            data = re_data[self.ticker][0][date_key].get(field_name)
        else:
            data = {}
            for tick in self.ticker:
                try:
                    date_key = re_data[tick][0].keys()[0]
                except:
                    try:
                        date_key = list(re_data[tick][0].keys())[0]
                    except:
                        date_key = None
                if date_key is not None:
                    sub_data = re_data[tick][0][date_key][field_name]
                    data.update({tick: sub_data})
                else:
                    data.update({tick: None})
        return data

    # Public method to get daily dividend data
    def get_daily_dividend_data(self, start_date, end_date):
        start = self.format_date(start_date)
        end = self.format_date(end_date)
        return self.get_stock_dividend_data(start, end, 'daily')

    # Public Price Data Methods
    def get_current_price(self):
        return self._stock_price_data('regularMarketPrice')

    def get_current_change(self):
        return self._stock_price_data('regularMarketChange')

    def get_current_percent_change(self):
        return self._stock_price_data('regularMarketChangePercent')

    def get_current_volume(self):
        return self._stock_price_data('regularMarketVolume')

    def get_prev_close_price(self):
        return self._stock_price_data('regularMarketPreviousClose')

    def get_open_price(self):
        return self._stock_price_data('regularMarketOpen')

    def get_ten_day_avg_daily_volume(self):
        return self._stock_summary_data('averageDailyVolume10Day')

    def get_stock_exchange(self):
        return self._stock_price_data('exchangeName')

    def get_market_cap(self):
        return self._stock_price_data('marketCap')

    def get_daily_low(self):
        return self._stock_price_data('regularMarketDayLow')

    def get_daily_high(self):
        return self._stock_price_data('regularMarketDayHigh')

    def get_currency(self):
        return self._stock_price_data('currency')

    # Public Summary Data Methods
    def get_yearly_high(self):
        return self._stock_summary_data('fiftyTwoWeekHigh')

    def get_yearly_low(self):
        return self._stock_summary_data('fiftyTwoWeekLow')

    def get_dividend_yield(self):
        return self._stock_summary_data('dividendYield')

    def get_annual_avg_div_yield(self):
        return self._stock_summary_data('trailingAnnualDividendYield')

    def get_five_yr_avg_div_yield(self):
        return self._stock_summary_data('fiveYearAvgDividendYield')

    def get_dividend_rate(self):
        return self._stock_summary_data('dividendRate')

    def get_annual_avg_div_rate(self):
        return self._stock_summary_data('trailingAnnualDividendRate')

    def get_50day_moving_avg(self):
        return self._stock_summary_data('fiftyDayAverage')

    def get_200day_moving_avg(self):
        return self._stock_summary_data('twoHundredDayAverage')

    def get_beta(self):
        return self._stock_summary_data('beta')

    def get_payout_ratio(self):
        return self._stock_summary_data('payoutRatio')

    def get_pe_ratio(self):
        return self._stock_summary_data('trailingPE')

    def get_price_to_sales(self):
        return self._stock_summary_data('priceToSalesTrailing12Months')

    def get_exdividend_date(self):
        return self._stock_summary_data('exDividendDate')

    # Financial Statement Data Methods
    def get_book_value(self):
        return self._financial_statement_data('balance', 'balanceSheetHistoryQuarterly',
                                              'totalStockholderEquity', 'quarterly')

    def get_ebit(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'ebit', 'annual')

    def get_net_income(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'netIncome', 'annual')

    def get_interest_expense(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'interestExpense', 'annual')

    def get_operating_income(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'netIncomeContinuousOperations',
                                              'annual')

    def get_total_operating_expense(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'totalOperatingExpenses', 'annual')

    def get_total_revenue(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'totalRevenue', 'annual')

    def get_cost_of_revenue(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'costOfRevenue', 'annual')

    def get_income_before_tax(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'incomeBeforeTax', 'annual')

    def get_income_tax_expense(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'incomeTaxExpense', 'annual')

    def get_gross_profit(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'grossProfit', 'annual')

    def get_net_income_from_continuing_ops(self):
        return self._financial_statement_data('income', 'incomeStatementHistory',
                                              'netIncomeFromContinuingOps', 'annual')

    def get_research_and_development(self):
        return self._financial_statement_data('income', 'incomeStatementHistory', 'researchDevelopment', 'annual')

    def get_recommendations(self):
        return self._get_analytic_data("recommendations")

    def get_insights(self):
        return self._get_analytic_data("insights")

    # Calculated Financial Methods
    def get_earnings_per_share(self):
        price_data = self.get_current_price()
        pe_ratio = self.get_pe_ratio()
        if isinstance(self.ticker, str):
            return eps(price_data, pe_ratio)
        else:
            ret_obj = {}
            for tick in self.ticker:
                re_val = eps(price_data[tick], pe_ratio[tick])
                ret_obj.update({tick: re_val})
            return ret_obj

    def get_num_shares_outstanding(self, price_type='current'):
        today_low = self._stock_summary_data('dayHigh')
        today_high = self._stock_summary_data('dayLow')
        cur_market_cap = self._stock_summary_data('marketCap')
        current = self.get_current_price()
        if isinstance(self.ticker, str):
            return num_shares_outstanding(cur_market_cap, today_low, today_high, price_type, current)
        else:
            ret_obj = {}
            for tick in self.ticker:
                re_data = num_shares_outstanding(cur_market_cap[tick], today_low[tick],
                                                 today_high[tick], price_type, current[tick])
                ret_obj.update({tick: re_data})
            return ret_obj
